File: sdk/python/c2cpo/key_manager.py
import os
import urllib.request
import json
import logging
from typing import Optional


class C2CPOKeyUnavailableError(Exception):
    """
    Raised when the master secret cannot be retrieved from any backend.

    C2CPO fails closed — if the key is unavailable, the encoder/decoder
    will not operate, preventing silent fallback to unprotected communication.
    """
    pass


class KeyManager:
    """
    Retrieves the C2CPO master secret from a configured backend.

    Supported backends (in priority order):

    1. **HashiCorp Vault** — set ``use_vault=True``, ``VAULT_ADDR``, ``VAULT_TOKEN``
    2. **Cloud KMS** — set ``C2CPO_KMS_BACKEND`` to ``AWS``, ``AZURE``, or ``GCP``
    3. **Environment variable** — set ``C2CPO_MASTER_SECRET`` (hex string, 64+ chars)

    The master secret is **never logged** at any level. Exceptions are wrapped
    to prevent accidental key material leakage in stack traces.

    Usage::

        # Simplest — environment variable
        export C2CPO_MASTER_SECRET="your-64-char-hex-string"

        key_manager = KeyManager()
        secret = key_manager.get_master_key()  # Returns bytes

        # HashiCorp Vault
        key_manager = KeyManager(use_vault=True)
        secret = key_manager.get_master_key()
    """

    def __init__(self, use_vault: bool = False):
        self.use_vault = use_vault
        self.vault_addr = os.getenv("VAULT_ADDR")
        self.vault_token = os.getenv("VAULT_TOKEN")

    def get_master_key(self) -> bytes:
        """
        Retrieve and return the master secret as bytes.

        Fails closed (raises C2CPOKeyUnavailableError) if the secret
        is unavailable from all configured backends.

        Returns:
            Master secret as bytes (minimum 32 bytes recommended).

        Raises:
            C2CPOKeyUnavailableError: If the secret cannot be retrieved.
        """
        try:
            if self.use_vault:
                key_hex = self._get_from_vault()
            elif os.getenv("C2CPO_KMS_BACKEND"):
                key_hex = self._get_from_cloud_kms()
            else:
                key_hex = os.getenv("C2CPO_MASTER_SECRET")

            if not key_hex:
                raise C2CPOKeyUnavailableError(
                    "Master secret not found in any configured backend. "
                    "Set C2CPO_MASTER_SECRET, VAULT_ADDR+VAULT_TOKEN, or C2CPO_KMS_BACKEND."
                )

            b_key = bytes.fromhex(key_hex)
            if len(b_key) < 32:
                logging.warning(
                    "C2CPO master secret is less than 32 bytes (256 bits). "
                    "Use a 64-character hex string (32 bytes) or longer for production."
                )
            return b_key

        except Exception as e:
            if isinstance(e, C2CPOKeyUnavailableError):
                raise
            raise C2CPOKeyUnavailableError(
                "Backend retrieval failed. Check logs for details."
            ) from e

    def _get_from_cloud_kms(self) -> Optional[str]:
        backend = os.getenv("C2CPO_KMS_BACKEND")

        if backend == "AWS":
            import boto3
            from botocore.exceptions import BotoCoreError, ClientError
            try:
                kms = boto3.client('kms')
                logging.info("Calling AWS KMS for C2CPO master secret.")
                return os.getenv("MOCK_AWS_KMS_RESPONSE")  # Replace with real KMS call
            except (BotoCoreError, ClientError) as e:
                raise C2CPOKeyUnavailableError("AWS KMS communication failed.") from e

        elif backend == "AZURE":
            logging.info("Retrieving from Azure Key Vault.")
            return os.getenv("MOCK_AZURE_KV_RESPONSE")  # Replace with real SDK call

        elif backend == "GCP":
            logging.info("Retrieving from GCP KMS.")
            return os.getenv("MOCK_GCP_KMS_RESPONSE")  # Replace with real SDK call

        return None

    def _get_from_vault(self) -> Optional[str]:
        if not self.vault_addr or not self.vault_token:
            raise C2CPOKeyUnavailableError(
                "VAULT_ADDR or VAULT_TOKEN not set in environment."
            )

        req = urllib.request.Request(
            f"{self.vault_addr}/v1/secret/c2cpo/master_key",
            headers={"X-Vault-Token": self.vault_token},
        )
        try:
            with urllib.request.urlopen(req, timeout=3.0) as response:
                body = json.loads(response.read().decode('utf-8'))
                return body.get("data", {}).get("key")
        except Exception as e:
            raise C2CPOKeyUnavailableError("Vault communication failed.") from e

File: sdk/nodejs/src/key_manager.ts
import * as http from 'http';
import * as https from 'https';

/**
 * Raised when the master secret cannot be retrieved from any backend.
 * C2CPO fails closed — if the key is unavailable, the encoder will not operate.
 */
export class C2CPOKeyUnavailableError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'C2CPOKeyUnavailableError';
  }
}

/**
 * Retrieves the C2CPO master secret from a configured backend.
 *
 * Supported backends (in priority order):
 * 1. **HashiCorp Vault** — set `useVault=true`, `VAULT_ADDR`, `VAULT_TOKEN`
 * 2. **Cloud KMS** — set `C2CPO_KMS_BACKEND` to `AWS`, `AZURE`, or `GCP`
 * 3. **Environment variable** — set `C2CPO_MASTER_SECRET` (hex string, 64+ chars)
 *
 * @example
 * ```typescript
 * // Simplest — environment variable
 * process.env.C2CPO_MASTER_SECRET = 'your-64-char-hex-string';
 * const secret = await new KeyManager().getMasterKey();
 *
 * // HashiCorp Vault
 * const secret = await new KeyManager(true).getMasterKey();
 * ```
 */
export class KeyManager {
  private useVault: boolean;
  private vaultAddr?: string;
  private vaultToken?: string;

  constructor(useVault: boolean = false) {
    this.useVault = useVault;
    this.vaultAddr = process.env.VAULT_ADDR;
    this.vaultToken = process.env.VAULT_TOKEN;
  }

  public async getMasterKey(): Promise<Buffer> {
    try {
      let keyHex: string | undefined;

      if (this.useVault) {
        keyHex = await this.getFromVault();
      } else if (process.env.C2CPO_KMS_BACKEND) {
        keyHex = await this.getFromCloudKms();
      } else {
        keyHex = process.env.C2CPO_MASTER_SECRET;
      }

      if (!keyHex) {
        throw new C2CPOKeyUnavailableError(
          'Master secret not found. Set C2CPO_MASTER_SECRET, VAULT_ADDR+VAULT_TOKEN, or C2CPO_KMS_BACKEND.'
        );
      }

      const bKey = Buffer.from(keyHex, 'hex');
      if (bKey.length < 32) {
        console.warn('C2CPO master secret is less than 32 bytes. Use a 64-char hex string for production.');
      }
      return bKey;
    } catch (e) {
      if (e instanceof C2CPOKeyUnavailableError) throw e;
      throw new C2CPOKeyUnavailableError(`Backend retrieval failed: ${e}`);
    }
  }

  private async getFromCloudKms(): Promise<string | undefined> {
    const backend = process.env.C2CPO_KMS_BACKEND;
    if (backend === 'AWS') {
      console.log('Calling AWS KMS for C2CPO master secret.');
      return process.env.MOCK_AWS_KMS_RESPONSE; // Replace with real AWS SDK call
    } else if (backend === 'AZURE') {
      console.log('Retrieving from Azure Key Vault.');
      return process.env.MOCK_AZURE_KV_RESPONSE; // Replace with real Azure SDK call
    } else if (backend === 'GCP') {
      console.log('Retrieving from GCP KMS.');
      return process.env.MOCK_GCP_KMS_RESPONSE; // Replace with real GCP SDK call
    }
    return undefined;
  }

  private async getFromVault(): Promise<string | undefined> {
    if (!this.vaultAddr || !this.vaultToken) {
      throw new C2CPOKeyUnavailableError('VAULT_ADDR or VAULT_TOKEN not set in environment.');
    }

    return new Promise((resolve, reject) => {
      const url = new URL(`${this.vaultAddr}/v1/secret/c2cpo/master_key`);
      const client = url.protocol === 'https:' ? https : http;

      const req = client.request(url, {
        method: 'GET',
        headers: { 'X-Vault-Token': this.vaultToken },
        timeout: 3000,
      }, (res) => {
        let data = '';
        res.on('data', (chunk) => data += chunk);
        res.on('end', () => {
          try {
            const body = JSON.parse(data);
            resolve(body?.data?.key);
          } catch (err) {
            reject(err);
          }
        });
      });

      req.on('error', reject);
      req.on('timeout', () => { req.destroy(); reject(new Error('Vault request timeout')); });
      req.end();
    });
  }
}
